pkg/vmprovider/providers/vsphere/virtualmachine/backup.go

```diff
@@ -90,22 +136,63 @@ func getEncodedVMKubeData(vm *vmopv1.VirtualMachine) (string, error) {
 	return util.EncodeGzipBase64(string(backupVMYaml))
 }
 
-func getEncodedVMBootstrapData(bootstrapData map[string]string) (string, error) {
-	if len(bootstrapData) == 0 {
+func constructVMObj(ecKubeData string) (vmopv1.VirtualMachine, error) {
+	var vmObj vmopv1.VirtualMachine
+	decodedKubeData, err := util.TryToDecodeBase64Gzip([]byte(ecKubeData))
+	if err != nil {
+		return vmObj, err
+	}
+
+	err = yaml.Unmarshal([]byte(decodedKubeData), &vmObj)
+	return vmObj, err
+}
+
+func getInstanceIDBackup(
+	vm *vmopv1.VirtualMachine,
+	ecMap map[string]string) (string, error) {
+	// Instance ID should not be changed once persisted in VM's ExtraConfig.
+	// Return an empty string to skip the backup if it already exists.
+	if _, ok := ecMap[constants.BackupVMInstanceIDExtraConfigKey]; ok {
 		return "", nil
 	}
 
-	bootstrapDataJSON, err := json.Marshal(bootstrapData)
+	instanceID := vm.Annotations[vmopv1.InstanceIDAnnotation]
+	if instanceID == "" {
+		instanceID = string(vm.UID)
+	}
+
+	return util.EncodeGzipBase64(instanceID)
```

Comment:
is it necessary to encode instanceID?

Reply:
It's nice to keep them all consistent so it's easier to parse when restoring the VM.